listener: support keepalive configuration

jronak opened the pull request:
Commit Message: This change introduces the ability to configure TCP keepalive settings on additional listener addresses. Currently users have to set the corresponding socket options directly and with this change users can enable TCP keepalive for both primary and additional addresses via the listener configuration similar to the cluster configuration.

Additional Description: Configuration example:
```
static_resources:
  listeners:
  - name: "listener"
    address:
      socket_address:
        address: 127.0.0.1
        port_value: 10000
    # TCP keepalive configuration for the primary address.
    tcp_keepalive:
      keepalive_time: 60000       # time in milliseconds before probing starts
      keepalive_interval: 10000   # interval between probes in milliseconds
      keepalive_probes: 5         # number of abandoned probes before closing connection
    additional_addresses:
      - address:
          socket_address:
            address: 127.0.0.2
            port_value: 10001
        # TCP keepalive configuration for the additional address.
        tcp_keepalive:
          keepalive_time: 60000
          keepalive_interval: 10000
          keepalive_probes: 5
```
Risk Level: low
Testing: unit and integration
Docs Changes: NA
Release Notes: NA

adisuissa reviewed (commented):
High-level comment:
prior to additional-addresses, the listener to address 1-1 mapping introduced some address related config knobs in the listener.
I think that in order to ensure that a coherent config design is used, the fields in the listener level will impact all the addresses (both the primary and additional), and the additional address will be able to override the config (just as being done for socket options).
I also think that it would be better to have a common-address-options (similar to UpstreamConnectionOptions) that can be used to optionally override these values.

adisuissa commented:
/wait

jronak commented:
Thanks for the comments @adisuissa 
> prior to additional-addresses, the listener to address 1-1 mapping introduced some address related config knobs in the listener.
I think that in order to ensure that a coherent config design is used, the fields in the listener level will impact all the addresses (both the primary and additional), and the additional address will be able to override the config (just as being done for socket options).

Updated the PR to use apply keepalive configuration across all the additional addresses and ability override/disable keepalive at each additional address level.

> I also think that it would be better to have a common-address-options (similar to UpstreamConnectionOptions) that can be used to optionally override these values.

Happy to update the PR to use common address options, am I understanding the definition correctly?

```
message AddressOptions {
  core.v3.TcpKeepalive tcp_keepalive ;
}

message AdditionalAddress {
  ...
  AddressOptions address_options_override;
}

message Listener {
  ...
  AddressOptions common_address_options;
} 
```

What about `socket_options` field?

wbpcode commented on api/envoy/config/listener/v3/listener.proto:
```diff
@@ -45,6 +52,13 @@ message AdditionalAddress {
   // or an empty list of :ref:`socket_options <envoy_v3_api_field_config.core.v3.SocketOptionsOverride.socket_options>`,
   // it means no socket option will apply.
   core.v3.SocketOptionsOverride socket_options = 2;
+
+  // Configures TCP keepalive settings for the additional address.
+  // If specified, this will override the listener :ref:`tcp_keepalive <envoy_v3_api_field_config.listener.v3.Listener.tcp_keepalive>` configuration.
+  // If the :ref:`tcp_keepalive <envoy_v3_api_field_config.listener.v3.AdditionalAddress.TcpKeepaliveOverride.tcp_keepalive>`
+  // within the override is not specified, it means TCP keepalive settings will not be enabled
+  // for the additional address.
+  TcpKeepaliveOverride tcp_keepalive_override = 3;
```
`SocketOptionsOverride` is necessary because we cannot determine if `repeated` field is explicitly set or not. So, we wrapped a `SocketOptionsOverride` message.

But for the TcpKeepalive, please just add a `core.v3.TcpKeepalive tcp_keepalive = 3;` here.

jronak replied:
I added the override for keepalive to provide a mechanism for disabling the keepalive configuration specifically at the additional address, even when it remains enabled at the listener level. For example, keepalive is enabled at the listener level, but explicitly disabled for one address:

```
name: listener
    address:
      socket_address: { address: 127.0.0.1, port_value: 1111 }
    tcp_keepalive: {} # enable keep alive for all addresses

    additional_addresses:
    - address:
        # keepalive is enable for this address by default
        socket_address: { address: 127.0.0.1, port_value: 2223 }
    - address:
        socket_address: { address: 127.0.0.1, port_value: 2222 }
        tcp_keepalive_override: {} # disable keepalive for this address
```

wbpcode reviewed (changes requested):
This new API will effect our LDS updating. Two addresses (with same ip/port and other socket options) with differen tcp keepalive should be treat as different addresses and should create different socket.

That means we should also enhance listener loading/updating processing to ensure it works as expected.

I will suggest you convert the `tcp_keepalive` to `socket_options` field when loading the protocol, because the LDS updating have handled the socket_options.

jronak commented:
> This new API will effect our LDS updating. Two addresses (with same ip/port and other socket options) with differen tcp keepalive should be treat as different addresses and should create different socket.
> 
> That means we should also enhance listener loading/updating processing to ensure it works as expected.
>
> I will suggest you convert the `tcp_keepalive` to `socket_options` field when loading the protocol, because the LDS updating have handled the socket_options.

Thanks for the review. I've updated the differ to include tcp keepalive settings while comparing socket options, and added a integration test which compares tcp-keepalive settings before and after lds update.

RyanTheOptimist commented:
Looks like this needs a main merge
/wait
